### PR TITLE
Page Templates: Track template insertion/removal caused by the template selector

### LINKED ISSUE #39357: Page Templates: Track template insertion/removal caused by the template selector
Looking at the tech implementation of the template selector, the template insertion has been switched in November to use `replaceInnerBlocks` instead of `insertBlocks` under the hood. Doing this switch in #37500, tracking wasn't considered and has been effectively removed with that change.

This means events are no longer being triggered for blocks inserted or removed as part of the template selector interactions. I think this should be fixed in order to be able to build reliable funnels like `block_inserted > block_published`. Currently, we only know about the "published" part.

I propose we add tracking for `replaceInnerBlocks` to fire the same events as regular `insertBlocks`/`removeBlocks` and we add a new property to the event to indicate the source of the event is the template selector.

### PR DESCRIPTION
**Pull request by marekhrabe**
#### Changes proposed in this Pull Request

* adds a filter in SPT so the tracking library can detect events being triggered during template insertion

#### Testing instructions

No need to checkout this PR. Just apply the diff below. 

* apply D38786-code to your sandbox (diff just for testing purposes, real deployment would be handled separately after merging this PR)
* pick a wpcom simple site with SPT active for testing
* sandbox that site.wordpress.com and widgets.wp.com
* on your (sandboxed) site, go to the editor and run `localStorage.debug = '*'` (it's easier to go through wp-admin for testing than calypso)
* refresh the page
* start a new page, you should see the layout selector and `wpcom-block-editor:tracking registering tracking handlers` message in console (you may need to use devtools' `filter` feature to filter your console logs by the term `track`)
* pick a layout
* watch dev console for messages, one of them should be `wpcom-block-editor:analytics:tracks Recording event "wpcom_block_inserted"` and in its props, you should see `from_template_selector: true`
* insert any other block manually to the page and there should be one more tracking message but *without* `from_template_selector`

Fixes #39357

### CODE REVIEW

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (retrofox) [inline apps/wpcom-block-editor/src/wpcom/features/tracking.js]: Wondering if does it worth defining a global constants file in order to share them across these plugins.

**Comment by obenland** [comment]
This looks pretty solid already. What are the next steps?

**Comment by marekhrabe** [comment]
> This looks pretty solid already. What are the next steps?

Testing instructions now added 🙌

**Comment by getdave** [comment]
Going to review this today.

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js
Comment (getdave) [inline apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js]: With `stubTrue` being the default value?

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js
Comment (getdave) [inline apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js]: These reformatting changes make it difficult to see what was added. However, given it was just `stubTrue` that's ok 👍 😄

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (getdave) [inline apps/wpcom-block-editor/src/wpcom/features/tracking.js]: So by default, this is `false` right? It's odd because when I inserted a Columns block the tracks event was missing `from_template_selector` prop.

Screenshot 2020-02-13 at 11 35 57

But then once the Block inserted it automatically added x2 `core/column` (singular!) Blocks and those did have the `from_template_selector` prop:

Screenshot 2020-02-13 at 11 40 01

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (getdave) [inline apps/wpcom-block-editor/src/wpcom/features/tracking.js]: Using a filter is really nice 👍

**Review by getdave — DISMISSED** [review]
This is looking really good.

I tested as per your (quite excellent) instructions and it all seemed to work perfectly.

The only query I have is around why when I manually add Blocks the `from_template_selector` prop appears to be either not present or set to `false`. Surely it should always be set to `false` no?

Other than that I'd say this is good to go (with the exception of getting ✅  on your tests!)

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (getdave) [reply apps/wpcom-block-editor/src/wpcom/features/tracking.js]: Also...just checking, if a filter isn't registered then I assume anything that hooks into that filter is ignored and doesn't throw an error. Just checking because otherwise if you don't ensure widgets is deployed in advance of this PR landing we're going to have errors.

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (marekhrabe) [reply apps/wpcom-block-editor/src/wpcom/features/tracking.js]: very WordPress!

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (marekhrabe) [reply apps/wpcom-block-editor/src/wpcom/features/tracking.js]: >  if a filter isn't registered then I assume anything that hooks into that filter is ignored

100% correct. it just returns the default value immediately, which is `false`

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js
Comment (marekhrabe) [reply apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js]: only `stubTrue` ✅

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js
Comment (marekhrabe) [reply apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js]: yeah, we need it to always return true and this is a convenience method for just that. we remove the filter after insertion so it will be back to `false`

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (marekhrabe) [reply apps/wpcom-block-editor/src/wpcom/features/tracking.js]: > Columns block the tracks event was missing from_template_selector prop.

When you insert a block manually, it uses a different tracking code path where `from_template_selector` can never be true so that's why the check is not included.

> Block inserted it automatically added x2 core/column (singular!) Blocks and those did have the from_template_selector prop:

It's using the same method template selector uses (`replaceInnerBlocks`) so the tracking will include `from_template_selector` and it should be `false` because no filter is registered at the moment to change it to anything else

Comment (getdave) [reply apps/wpcom-block-editor/src/wpcom/features/tracking.js]: Thanks for thorough explanations.

**Comment by retrofox** [comment]
LGTM. Tested in my sandbox and tracks when the blocks are inserted from the template selector:

image

question: it doesn't track blocks contained into other blocks. For instance, the `about` template...
image

...has many blocks which are contained by the `core/group block`. Are we going to ignore tracking these blocks? Make sense of tracking them?

In fact, on this specific example, when the template is inserted we are tracking that only `core/group` blocks are been inserted, but the template is full of other ones such as image, paragraph, columns, etc

**Comment by getdave** [comment]
Screenshot 2020-02-18 at 10 35 24

Same concern here @retrofox. I see only `core/group` events being tracked. I'm going to go through code again to get a fuller understanding and see if I can use this to inform what is going on.

**Comment by marekhrabe** [comment]
You are correct, I was about to make an issue for recursive tracking as I have noticed it too with my testing. I would say it's out of scope for this change, rather a new feature that we need to plan for iteration. Would you agree?

**Comment by retrofox** [comment]
> You are correct, I was about to make an issue for recursive tracking as I have noticed it too with my testing. I would say it's out of scope for this change, rather a new feature that we need to plan for iteration. Would you agree?

Honestly, to me, it's like a blocker since the implementation doesn't tackle the main purpose of the PR. 
Also, populating partially the stats registering only `core/group` events and ignoring the other blocks seem to be something not very well.
Anyway, it's totally up to you. We can continue iterating immediately after merging this one. 😃👍

**Comment by getdave** [comment]
> Honestly, to me, it's like a blocker since the implementation doesn't tackle the main purpose of the PR

I am on the fence here. Initially I agreed with @retrofox but now remember my recent troubles, I'm inclined towards merging what we have and following up with incremental improvements.

Tracking something might be better than nothing unless it significantly skews the event data to make it look like `core/group` is the best block ever (which it is of course! 😉).

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (getdave) [inline apps/wpcom-block-editor/src/wpcom/features/tracking.js]: If before `castArray()` `blocks` is a deeply nested object then presumably we can recurse into it and call the same callback you provide to `forEach` to handle dispatching events for _all_ blocks and not just the top level?

Comment (getdave) [reply apps/wpcom-block-editor/src/wpcom/features/tracking.js]: Just checked and `blocks` is

> Block objects to insert as new InnerBlocks

To presumably it is nested.

Comment (marekhrabe) [reply apps/wpcom-block-editor/src/wpcom/features/tracking.js]: We can flatten the blocks here as you suggest. I haven't done it immediately as I was wondering if other actions we are tracking should also be looked at recursively, hence the suggestion to make it a separate task and get answers first.

Comment (retrofox) [reply apps/wpcom-block-editor/src/wpcom/features/tracking.js]: >  I was wondering if other actions we are tracking should also be looked at recursively

Yes, it's a very good point, too.

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (getdave) [inline apps/wpcom-block-editor/src/wpcom/features/tracking.js]: Yep this is the recursion that we need 👍

**Review by getdave — COMMENTED** [review]
I tested this with @retrofox's changes to add recursion and it appears to be working in that we now have a large number of Blocks being tracked whereas previously it was only tracking the top-level `core/group`.

I feel like this is an improvement over the original as we are now actually tracking all the blocks that get inserted and not just the topmost ones.

We should probably be careful about taking this any further. We should consider shipping "as is" and following up with further refinements in future PRs.

Great work @marekhrabe and @retrofox!

### Testing Details
Screenshot 2020-02-20 at 11 54 05

I chose this template for testing:

Screenshot 2020-02-20 at 11 57 05

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (getdave) [inline apps/wpcom-block-editor/src/wpcom/features/tracking.js]: I see these are args to the callback function within the timeout right?

**Comment by retrofox** [comment]
I've updated added some changes in order to track all blocks, and it populates the event with `inner_block`, `parent_block_client_id` and `parent_block_name`

image

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (getdave) [reply apps/wpcom-block-editor/src/wpcom/features/tracking.js]: Maybe you are trying to defer the calls to avoid blocking the editor from rendering? If so then this will push each call onto the next "tick" of the Event loop:

```js
setTimeout(fn, 0)
```

Guess it's much the same. Either way we should comment this to make sure it's clear what's going on.

> This method can be used to break up long running operations and run a callback function immediately after the browser has completed other operations such as events and display updates.

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (retrofox) [reply apps/wpcom-block-editor/src/wpcom/features/tracking.js]: yeah, it ensures that we are using the correct var values in the function scope.

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (getdave) [inline apps/wpcom-block-editor/src/wpcom/features/tracking.js]: I was wondering about this. I think it's good to have this info.

Comment (retrofox) [reply apps/wpcom-block-editor/src/wpcom/features/tracking.js]: yeah, it's related whit the @marekhrabe comment

>  ... haven't done it immediately as I was wondering if other actions we are tracking should also be looked at recursively,...

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (marekhrabe) [inline apps/wpcom-block-editor/src/wpcom/features/tracking.js]: I think this should be true when doing a block replacement.

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (retrofox) [reply apps/wpcom-block-editor/src/wpcom/features/tracking.js]: Yeah.
First of all, probably it doesn't make sense.

One reason to delay tracking these events is avoiding to affect with the template rendering process. A template could contain hundreds of blocks and waiting to track all events could affect it in terms of performance. Again, not sure about this.
My other concern is triggering these requests at the same time. For this reason, I've delayed, one by one, every tacking event. I think that 200 requests at the same time could be something undesired, but definitely it worths research and confirm this fragile theory.
Anyway, removing this will get the approach easier and simpler. Glad to do it.

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (marekhrabe) [reply apps/wpcom-block-editor/src/wpcom/features/tracking.js]: I'm not sure the client_id would be useful. These ids are an implementation detail of the block editor and they are not persistent across edit sessions so I don't think we could gain anything from having them. The block name though is a perfect idea!

**Comment by marekhrabe** [comment]
The functionality looks good but it doesn't feel right with the `setTimeout` and a module-global variable added. It feels strange to solve the concurrency problem on this level. Did you encounter any serious perf issues without it or did it just felt wrong in general to do all of those requests at once? 

If it's not already critical, I would suggest removing the `setTimeout` from here and calling `tracksRecordEvent` immediately in the loop. If we encounter performance problems, I would update the `tracksRecordEvent` function with some limitation logic so all tracking calls would get the benefit, not just this single loop.

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (retrofox) [reply apps/wpcom-block-editor/src/wpcom/features/tracking.js]: does it mean that `blocks_replaced` should be true only for the root blocks? So for the inner blocks it should be false?

**Comment by retrofox** [comment]
> The functionality looks good but it doesn't feel right with the `setTimeout` and a module-global variable added. It feels strange to solve the concurrency problem on this level.

Yeah, far enough.

> Did you encounter any serious perf issues without it or did it just felt wrong in general to do all of those requests at once?

No, I don't. This implementation is based on a simple assumption of mine.


> If it's not already critical, I would suggest removing the `setTimeout` from here and calling `tracksRecordEvent` immediately in the loop. If we encounter performance problems, I would update the `tracksRecordEvent` function with some limitation logic so all tracking calls would get the benefit, not just this single loop.

👍
Thanks

**Comment by marekhrabe** [comment]
I think this is ready now, awaiting one last approval after all new changes introduced.

For the shipping, we need to publish a new release of the Full Site Editing Plugin PCYsg-mMA-p2. After that is done (the order is important), we need to update `wpcom-block-editor` as per PCYsg-l4k-p2

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (marekhrabe) [reply apps/wpcom-block-editor/src/wpcom/features/tracking.js]: If the newly inserted block replaced anything, it should be true. I think inner blocks should inherit it, as they are part of the replacement.

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js
Comment (obenland) [inline apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js]: Should we define these in a separate file so they can be imported in `tracking.js`?

Comment (retrofox) [reply apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js]: I made the same suggestion at the beginning, but in some way, it's like in WordPress filters, right? You need to know the exact name of the filter in order to get rightly hooked.
Anyway, I agree with adding a global file definition, too.

Comment (marekhrabe) [reply apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js]: Since FSE and wpcom-block-editor are completely separate apps and I believe there is no precedent of any shared code other than npm-based libraries, I think we should keep them separate. In WordPress backend you also write all the filter names directly in files and you don't use a central place with a list of filters to import.

**Comment by obenland** [comment]
@marekhrabe It looks like that phab diff is out of date. Was there a way to test the tracking locally by setting a variable in the console?

**Comment by getdave** [comment]
Testing now

**Review by getdave — APPROVED** [review]
I tested this on WPCom using WPAdmin. 

I selected the "About" template:

Screenshot 2020-02-25 at 10 51 43

...and I saw these events in console when filtering by the term `track`:

Screenshot 2020-02-25 at 10 51 12

These appear to marry up well with that is in the Template. I believe the recursive tracking is working.

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (getdave) [inline apps/wpcom-block-editor/src/wpcom/features/tracking.js]:
```suggestion
 * in order to track inner blocks.
```

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (getdave) [inline apps/wpcom-block-editor/src/wpcom/features/tracking.js]:
```suggestion
 * Also, it adds default `inner_block`,
```

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js
Comment (getdave) [inline apps/wpcom-block-editor/src/wpcom/features/tracking.js]:
```suggestion
 * Page Templates insert their content into the page replacing everything that was already there.
```

**Review by getdave — COMMENTED** [review]
Some tiny minor comment wording tweaks I forgot on my other review.

**Comment by marekhrabe** [comment]
Accepted all of them, thanks a lot :)

**Pull request merged**
